File: 0155-min-stack/0155-min-stack.py
class MinStack(object):
    def __init__(self):
        self.stack = []
        self.minv = []

    def push(self, val):
        self.stack.append(val)
        if not self.minv:
            self.minv.append(val)
        if val < self.minv[0]:
            self.minv[0] = val

    def pop(self):
        self.stack.pop(-1)
        # del 함수의 time complexity: O(n)

    def top(self):
        return self.stack[-1]

    def getMin(self):
        # return min(self.stack)
        # min 메서드의 time complexity: O(n)
        return self.minv[0]



# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: 0155-min-stack/0155-min-stack.cpp
#include <stack>
class MinStack {
public:
    stack<int> st;
    stack<int> minv;
    MinStack() {    }
    
    void push(int val) {
        st.push(val);
        if (minv.empty()) minv.push(val);
        else if (minv.top() > val) minv.push(val);
        else minv.push(minv.top());
    }
    
    void pop() {
        st.pop();
        minv.pop();
    }
    
    int top() {
        return st.top();
    }
    
    int getMin() {
        return minv.top();
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */